File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque
def solution(maps):
    dx = [0,0,1,-1]
    dy = [1,-1,0,0]
    n = len(maps)
    m = len(maps[0])
    queue = deque([[0,0]])
    
    while queue:
        x, y = queue.popleft()
        for i in range(4) :
            nx = x + dx[i]
            ny = y + dy[i]
            if nx > n-1 or ny > m-1 or nx < 0 or ny < 0 :
                continue
            if (maps[nx][ny] == 1 ) :
                maps[nx][ny] = maps[x][y] + 1
                
                queue.append([nx,ny])
    
    answer = maps[n-1][m-1]
    if (answer == 1) :
        return -1
    return answer

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    public int solution(int[][] maps) {
        int answer = 0;
        int[] dx = {0, 0, 1, -1};
        int[] dy = {1, -1, 0, 0};
        int n = maps.length;
        int m = maps[0].length;
        Queue<List<Integer>> queue = new LinkedList<>();
        queue.offer(List.of(0, 0));
        
        while(!queue.isEmpty()) {
            List<Integer> xy = queue.poll();
            Integer x = xy.get(0);
            Integer y = xy.get(1);
            for (int i = 0; i < 4; i++) {
                Integer nx = x + dx[i];
                Integer ny = y + dy[i];
                if (nx < 0 || nx > n-1 || ny < 0 || ny > m-1) {
                    continue;
                }
                
                if (maps[nx][ny] == 1) {
                    maps[nx][ny] = maps[x][y] +1;
                    queue.offer(List.of(nx, ny));
                }
                
            }
        }
        
        return maps[n-1][m-1] == 1 ? -1 : maps[n-1][m-1];
    }
}